File: esm-order-review-main (6)/esm-order-review-main/ESM_SCM8/backend/models/support.py
"""Persistent customer-support ticket models."""

from __future__ import annotations

from datetime import date, datetime, timezone
from uuid import uuid4

from sqlalchemy import BigInteger, Date, DateTime, ForeignKey, Index, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from backend.database import Base


def utcnow() -> datetime:
    return datetime.now(timezone.utc)


class SupportTicket(Base):
    __tablename__ = "support_tickets"
    __table_args__ = (
        Index("ix_support_tickets_status_created_at", "status", "created_at"),
        Index("ix_support_tickets_requester_created_at", "requester_name", "created_at"),
        Index("ix_support_tickets_owner_created_at", "owner_username", "created_at"),
    )

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid4()))
    ticket_number: Mapped[str] = mapped_column(String(32), unique=True, nullable=False, index=True)
    # Nullable only for tickets created before account ownership was introduced.
    # Legacy rows stay visible to administrators but are never guessed or
    # automatically assigned from the free-form requester name.
    owner_username: Mapped[str | None] = mapped_column(String(100))
    requester_name: Mapped[str] = mapped_column(String(100), nullable=False)
    requester_team: Mapped[str | None] = mapped_column(String(100))
    category: Mapped[str] = mapped_column(String(40), nullable=False)
    title: Mapped[str] = mapped_column(String(200), nullable=False)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="접수")
    assignee: Mapped[str | None] = mapped_column(String(100))
    due_date: Mapped[date | None] = mapped_column(Date)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utcnow, nullable=False)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utcnow, onupdate=utcnow, nullable=False)

    attachments: Mapped[list[SupportAttachment]] = relationship(
        back_populates="ticket", cascade="all, delete-orphan"
    )
    status_events: Mapped[list[SupportStatusEvent]] = relationship(
        back_populates="ticket", cascade="all, delete-orphan"
    )


class SupportAttachment(Base):
    __tablename__ = "support_attachments"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid4()))
    ticket_id: Mapped[str] = mapped_column(
        ForeignKey("support_tickets.id", ondelete="CASCADE"), nullable=False, index=True
    )
    original_name: Mapped[str] = mapped_column(String(255), nullable=False)
    storage_key: Mapped[str] = mapped_column(String(500), nullable=False, unique=True)
    content_type: Mapped[str | None] = mapped_column(String(120))
    size_bytes: Mapped[int] = mapped_column(BigInteger, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utcnow, nullable=False)

    ticket: Mapped[SupportTicket] = relationship(back_populates="attachments")


class SupportStatusEvent(Base):
    __tablename__ = "support_status_events"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid4()))
    ticket_id: Mapped[str] = mapped_column(
        ForeignKey("support_tickets.id", ondelete="CASCADE"), nullable=False, index=True
    )
    from_status: Mapped[str | None] = mapped_column(String(20))
    to_status: Mapped[str] = mapped_column(String(20), nullable=False)
    changed_by: Mapped[str] = mapped_column(String(100), nullable=False)
    note: Mapped[str | None] = mapped_column(Text)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utcnow, nullable=False)

    ticket: Mapped[SupportTicket] = relationship(back_populates="status_events")

File: esm-order-review-main (6)/esm-order-review-main/ESM_SCM8/frontend/lib/api/support.ts
import { apiFetch, ApiRequestError, FASTAPI_BASE_URL, parseApiError } from "./client";

export type SupportAttachment = {
  id: string;
  original_name: string;
  content_type: string | null;
  size_bytes: number;
};

export type SupportTicket = {
  id: string;
  ticket_number: string;
  requester_name: string;
  requester_team: string | null;
  category: string;
  title: string;
  content: string;
  status: string;
  assignee: string | null;
  due_date: string | null;
  created_at: string;
  updated_at: string;
  attachments: SupportAttachment[];
};

export type SupportTicketSummary = {
  total: number;
  received: number;
  in_progress: number;
  completed: number;
  delayed: number;
};

export type SupportTicketList = {
  items: SupportTicket[];
  summary: SupportTicketSummary;
};

export type CreateSupportTicketInput = {
  requesterName: string;
  requesterTeam?: string;
  category: string;
  title: string;
  content: string;
  attachment?: File | null;
};

export async function getSupportTickets(): Promise<SupportTicketList> {
  const response = await apiFetch(`${FASTAPI_BASE_URL}/support/tickets`, {
    method: "GET",
    cache: "no-store"
  });
  if (!response.ok) {
    throw new ApiRequestError(await parseApiError(response), response.status);
  }
  return response.json() as Promise<SupportTicketList>;
}

export async function createSupportTicket(input: CreateSupportTicketInput): Promise<SupportTicket> {
  const form = new FormData();
  form.set("requester_name", input.requesterName);
  if (input.requesterTeam) form.set("requester_team", input.requesterTeam);
  form.set("category", input.category);
  form.set("title", input.title);
  form.set("content", input.content);
  if (input.attachment) form.set("attachment", input.attachment);

  const response = await apiFetch(`${FASTAPI_BASE_URL}/support/tickets`, {
    method: "POST",
    body: form
  });
  if (!response.ok) {
    throw new ApiRequestError(await parseApiError(response), response.status);
  }
  return response.json() as Promise<SupportTicket>;
}
